src/OpenTelemetry/Logs/LogRecordData.cs

```diff
@@ -0,0 +1,130 @@
+// <copyright file="LogRecordData.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace OpenTelemetry.Logs
+{
+    /// <summary>
+    /// Stores details about a log message.
+    /// </summary>
+    internal struct LogRecordData
+    {
+        internal DateTime TimestampBacking = DateTime.UtcNow;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LogRecordData"/> struct.
+        /// </summary>
+        /// <remarks>
+        /// Note: The <see cref="Timestamp"/> property is initialized to <see
+        /// cref="DateTime.UtcNow"/> automatically.
+        /// </remarks>
+        /// <param name="activity">Optional <see cref="Activity"/> used to populate context fields.</param>
+        public LogRecordData(Activity? activity = null)
+        {
+            if (activity != null)
```

Comment:
Can the ctor here use the static method `SetActivityContext` to avoid code duplication?

Reply:
Good question! Sadly, no. This is a weird quick of `struct`s. The compiler doesn't let you access `this` until the `struct` has been fully assigned. The reason to call the helper would be to do that assignment so it violates the rule.

```csharp
        public LogRecordData(Activity? activity = null)
        {
            ref LogRecordData instance = ref this; // Error CS0188  The 'this' object cannot be used before all of its fields have been assigned

            SetActivityContext(ref instance, activity);
        }
```

Reply:
i missed that this was a struct. thanks for the explanation!